drivers/dma/dma_stm32.c

```diff
@@ -0,0 +1,617 @@
+		config->head_block->dest_reload_en) {
+		LOG_ERR("source_reload_en and dest_reload_en must "
+			"be the same.");
+		return -EINVAL;
+	}
+
+	stream->busy		= true;
+	stream->dma_callback	= config->dma_callback;
+	stream->direction	= config->channel_direction;
+	stream->callback_arg    = config->callback_arg;
+
+	if (stream->direction == MEMORY_TO_PERIPHERAL) {
+		DMA_InitStruct.MemoryOrM2MDstAddress =
+					config->head_block->source_address;
+		DMA_InitStruct.PeriphOrM2MSrcAddress =
+					config->head_block->dest_address;
+	} else {
+		DMA_InitStruct.PeriphOrM2MSrcAddress =
+					config->head_block->source_address;
+		DMA_InitStruct.MemoryOrM2MDstAddress =
+					config->head_block->dest_address;
+	}
+
+	u16_t memory_addr_adj, periph_addr_adj;
+
+	ret = dma_stm32_get_priority(config->channel_priority,
+				     &DMA_InitStruct.Priority);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = dma_stm32_get_direction(config->channel_direction,
+				      &DMA_InitStruct.Direction);
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (config->channel_direction) {
+	case MEMORY_TO_MEMORY:
+	case PERIPHERAL_TO_MEMORY:
+		memory_addr_adj = config->head_block->dest_addr_adj;
+		periph_addr_adj = config->head_block->source_addr_adj;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		memory_addr_adj = config->head_block->source_addr_adj;
+		periph_addr_adj = config->head_block->dest_addr_adj;
+		break;
+	/* Direction has been asserted in dma_stm32_get_direction. */
+	}
+
+	ret = dma_stm32_get_memory_increment(memory_addr_adj,
+					&DMA_InitStruct.MemoryOrM2MDstIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+	ret = dma_stm32_get_periph_increment(periph_addr_adj,
+					&DMA_InitStruct.PeriphOrM2MSrcIncMode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->head_block->source_reload_en) {
+		DMA_InitStruct.Mode = LL_DMA_MODE_CIRCULAR;
+	} else {
+		DMA_InitStruct.Mode = LL_DMA_MODE_NORMAL;
+	}
+
+	source_periph = stream->direction == MEMORY_TO_PERIPHERAL;
+	ret = dma_stm32_width_config(config, source_periph, dma,
+				     &DMA_InitStruct, id);
+	if (ret < 0) {
+		return ret;
+	}
+	msize = DMA_InitStruct.MemoryOrM2MDstDataSize;
+
+#if defined(CONFIG_DMA_STM32_V1)
+	DMA_InitStruct.MemBurst = stm32_dma_get_mburst(config, source_periph);
+	DMA_InitStruct.PeriphBurst = stm32_dma_get_pburst(config,
+							source_periph);
+
+	if (config->channel_direction != MEMORY_TO_MEMORY) {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot error.");
+			return -EINVAL;
+		}
+	} else {
+		if (config->dma_slot >= 8) {
+			LOG_ERR("dma slot is too big, using 0 as default.");
+			config->dma_slot = 0;
+		}
+	}
+	DMA_InitStruct.Channel = table_ll_channel[config->dma_slot];
+
+	stm32_dma_get_fifo_threshold(config->head_block->fifo_mode_control);
+
+	if (stm32_dma_check_fifo_mburst(&DMA_InitStruct)) {
+		DMA_InitStruct.FIFOMode = LL_DMA_FIFOMODE_ENABLE;
+	} else {
+		DMA_InitStruct.FIFOMode = LL_DMA_FIFOMODE_DISABLE;
+	}
+#endif
+	DMA_InitStruct.NbData = config->head_block->block_size;
```

Comment:
Block size documentation states: `block_size is the number of bytes to be transferred for this block.`
Datasheet on NbData states: `NDT[15:0]: number of data items to transfer (0 up to 65535)` 

So I've tested this, and it happens to be that there is no isr fired when I' m using word sized transfers. 
Only the half transfer interrupt flag is set by the DMA controller.

```suggestion
	DMA_InitStruct.NbData = config->head_block->block_size / config->source_data_size;
```

Reply:
Thanks, I think you're right on this. I've also tested this this morning. This error would cause the DMA controller to transfer more data than we expected.
In your test environment I think the problem is that with word size, the DMA controller would be trying to access some address outside our src/dst data structures and it maybe illegle, so the program got stuck.

Reply:
According to the Reference Manual (Which I'm referencing now is 9.3.10 of RM0368, STM32F401xB), When the source data size and dest data size is different, the NDT is determined by PSIZE. So we should fix this to  
```suggestion
if (source_periph) {
	DMA_InitStruct.NbData = config->head_block->block_size / config->source_data_size;
} else {
	DMA_InitStruct.NbData = config->head_block->block_size / config->dest_data_size;
}
```